main.py

```diff
@@ -197,72 +225,112 @@ def cost_fn(params, masked_circuit=None):
+        )
+        costs.append(
+            cost_basis(
+                circuit,
                 masked_circuit.differentiable_parameters,
                 current_data,
+                current_target,
                 rotations,
                 masked_circuit,
                 masked_circuit.wires,
                 wires_to_measure,
-            )
-            costs.append(
-                cost_basis(
-                    circuit,
-                    masked_circuit.differentiable_parameters,
-                    current_data,
-                    current_target,
-                    rotations,
-                    masked_circuit,
-                    masked_circuit.wires,
-                    wires_to_measure,
-                    interpret,
-                )
-            )
-            selected_output = output[
                 interpret,
-            ]
-            same = np.argmax(current_target) == np.argmax(selected_output)
-            if same:
-                correct += 1
-            if __debug__:
-                print(
-                    f"Label: {current_target} Output: {selected_output} "
-                    f"({output}) Correct: {same}"
-                )
-
-        accuracy = correct / N
-        loss = np.average(costs)
+            )
+        )
+        selected_output = output[
+            interpret,
+        ]
+        same = np.argmax(current_target) == np.argmax(selected_output)
+        if same:
+            correct += 1
         if __debug__:
             print(
-                f"Accuracy = {correct} / {N} = {accuracy} \n",
-                f"Avg Cost: {loss}",
+                f"Label: {current_target} Output: {selected_output} "
+                f"({output}) Correct: {same}"
             )
-        return {
-            "accuracy": accuracy,
-            "loss": loss,
-        }
+    accuracy = correct / count
+    loss = np.average(costs)
+    if __debug__:
+        print(
+            f"[{variant}] Accuracy = {correct} / {count} = {accuracy}\n",
+            f"[{variant}] Avg Cost: {loss}",
+        )
+    return {
+        "accuracy": accuracy,
+        "loss": loss,
+    }
+
+
+def validate(
+    circuit,
+    masked_circuit: MaskedCircuit,
+    rotations: List,
+    wires_to_measure: Tuple[int, ...],
+    interpret: Tuple[int, ...],
+    data: np.ndarray,
+    target: np.ndarray,
+):
+    return loss_and_accuracy(
+        circuit,
+        masked_circuit=masked_circuit,
+        rotations=rotations,
+        wires_to_measure=wires_to_measure,
+        interpret=interpret,
+        data=data,
+        target=target,
+        variant="validate",
+    )
+
+
+def test(
+    circuit,
+    masked_circuit: MaskedCircuit,
+    rotations: List,
+    wires_to_measure: Tuple[int, ...],
+    interpret: Tuple[int, ...],
+    data: Optional[np.ndarray] = None,
+    target: Optional[np.ndarray] = None,
+):
+    if data is None or target is None:
+        pass
+    elif data is not None and target is not None:
```

Comment:
The ``pass`` case is already excluded by this one. You can remove it.
```suggestion
    if data is not None and target is not None:
```